libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1947,86 +2014,174 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+								"Invalid type for argument in function call. "
+								"Invalid implicit conversion from " +
+								type(*arguments[i])->toString() +
+								" to " +
+								parameterTypes[i]->toString() +
+								" requested."
+							);
+						break;
 					}
 
-			// check actual types
-			if (!duplication)
-				for (size_t i = 0; i < argumentNames.size(); i++)
-				{
-					bool found = false;
-					for (size_t j = 0; j < parameterNames.size(); j++)
-						if (parameterNames[j] == *argumentNames[i])
-						{
-							found = true;
-							// check type convertible
-							if (!type(*arguments[i])->isImplicitlyConvertibleTo(*parameterTypes[j]))
-								m_errorReporter.typeError(
-									arguments[i]->location(),
-									"Invalid type for argument in function call. "
-									"Invalid implicit conversion from " +
-									type(*arguments[i])->toString() +
-									" to " +
-									parameterTypes[i]->toString() +
-									" requested."
-								);
-							break;
-						}
-
-					if (!found)
-						m_errorReporter.typeError(
-							_functionCall.location(),
-							"Named argument \"" + *argumentNames[i] +  "\" does not match function declaration."
-						);
-				}
-		}
+				if (!found)
+					m_errorReporter.typeError(
+						_functionCall.location(),
+						"Named argument \"" +
+						*argumentNames[i] +
+						"\" does not match function declaration."
+					);
+			}
+
+		return;
 	}
+}
 
-	if (returnTypes.size() == 1)
-		_functionCall.annotation().type = returnTypes.front();
+bool TypeChecker::visit(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+	bool argumentsArePure = true;
+
+	// We need to check arguments' type first as they will be needed for overload resolution.
+	for (ASTPointer<Expression const> const& argument: arguments)
+	{
+		argument->accept(*this);
+		if (!argument->annotation().isPure)
+			argumentsArePure = false;
+	}
+
+	// For positional calls only, store argument types
+	if (_functionCall.names().empty())
+	{
+		shared_ptr<TypePointers> const argumentTypes = make_shared<TypePointers>();
+		for (ASTPointer<Expression const> const& argument: arguments)
+			argumentTypes->push_back(type(*argument));
+		_functionCall.expression().annotation().argumentTypes = move(argumentTypes);
+	}
+
+	_functionCall.expression().accept(*this);
+	TypePointer const expressionType = type(_functionCall.expression());
+
+	if (auto const* const typeType = dynamic_cast<TypeType const*>(expressionType.get()))
+	{
+		auto const actualType = typeType->actualType();
+
+		if (actualType->category() == Type::Category::Struct)
+		{
+			FunctionTypePointer const functionType =
+				dynamic_cast<StructType const&>(*actualType).constructorType();
+			_functionCall.annotation().isPure = argumentsArePure;
+			_functionCall.annotation().kind = FunctionCallKind::StructConstructorCall;
+			_functionCall.annotation().type = typeCheckFunctionLikeAndRetrieveReturnType(
```

Comment:
Shouldn't "struct constructor" be its own thing?

Reply:
I'd say it's debatable, at type checking level constructors and functions are syntically very similar - and the logic applied to constructors is more similar to vanilla functions than some of the special function types (i.e. ABI decode, etc.).

In general, my preference is for less code and more data driven validation - i.e. validating number and types of arguments against parameter definitions in the function type object rather than handling each function type kind specifically; for example, how the four BareCall and friends are treated to append extra text to the message.

Of course custom error messages may be considered necessary despite the increased ad-hoc logic - as this is not for me judge, I purposefully tried to ensure that all generated error messages were unchanged.